docs/website/contents/for-developers/NodeTasks.md

```diff
@@ -0,0 +1,303 @@
+
+      AddBlockRunner -- maintains --> CurrentSelection
+    end
+
+    ChainSelection -- "backed by" --> StorageLayer
+
+    subgraph StorageLayer[StorageLayer]
+      direction TB
+
+      VolatileDB["`**VolatileDB**`"]
+      ImmutableDB["`**ImmutableDB**`"]
+      LedgerDB["`**LedgerDB**`"]
+
+      CopyToImmutableDB{{CopyToImmutableDB}}
+      GarbageCollect{{GarbageCollect}}
+      CreateLedgerSnapshot{{CreateLedgerSnapshot}}
+
+      VolatileDB --> CopyToImmutableDB
+      CopyToImmutableDB --> ImmutableDB
+      GarbageCollect -- "cleans up" --> VolatileDB
+      LedgerDB --> CreateLedgerSnapshot
+    end
+  end
+
+  UpstreamPeersAndMint["Upstream peers and minting"]
+
+  UpstreamPeersAndMint -- "new blocks" --> ChainSelection
+```
+
+ - **AddBlockRunner**:
+   Sequentially process incoming blocks (from upstream peers, or minted by this node if it is a block producer).
+   A new block is first stored in the VolatileDB, and might become part of the selection (only if it is valid):
+
+    - Common case: The block fits on top of the current selection.
+    - The block is part of a fork that is preferrable to the current selection.
+      Usually, this is the case when the block has the same predecessor as the tip block of the selection, and a better tiebreaker.
+ - **CreateLedgerSnapshot**:
+   From time to time (by default, every 72 min for a caught-up node), the node writes the ledger state to disk.
+   This is a rather expensive operation as the ledger state is quite large.[^snapshots-slow]
+ - **CopyToImmutableDB**:
+   Once a block on the selection is confirmed by sufficiently many blocks, it becomes immutable and is copied from the VolatileDB to the ImmutableDB.
+   This is a very cheap operation.
+ - **GarbageCollect**:
+   Blocks that are either immutable or can no longer be part of a fork we might switch to are occasionally deleted from the VolatileDB.
+   This is a very cheap operation.
+
+### The Mempool
+
+The mempool maintains a bounded sequence of transactions that are valid when applied to the ledger state corresponding to the tip of the current selection.
+When the node is a block producer, this is the source of the contents of its minted blocks.
+
+```mermaid
+graph LR
+  subgraph MempoolComponent[Mempool]
+    Mempool[Mempool]
+    MempoolSyncState{{MempoolSyncState}}
+    MempoolSyncState -- updates --> Mempool
+  end
+
+  LedgerDB
+  LedgerDB -- "wake up when the tip\nof the selection changes" --> MempoolSyncState
+
+  DownstreamPeers
+  DownstreamPeers -- "new txs" --> Mempool
+
+  LocalClients
+  LocalClients -- "new txs" --> Mempool
+```
+
+ - **Adding new transactions.**
+   New transactions can be added via downstream peers (another node) or local clients (like cardano-cli).
+   The transaction is only added to the mempool if it is fully valid, and otherwise discarded.[^tx-validity]
+
+   Adding a transaction requires the acquisition of a lock, so no transactions are validated in parallel.
```

Comment:
I might be mistaken, but I think this is not true on pre-utxo-hd nodes. I think the mempool is an STM component.

Reply:
The lock is acquired here:
https://github.com/IntersectMBO/ouroboros-consensus/blob/85f73484829f4808f62b3bb9bd465a03b6d04edb/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Mempool/Update.hs#L73-L86

Reply:
Ah indeed